src/unix/filesystem.h

```diff
@@ -52,6 +52,35 @@ static inline const char *filename_from_path(const char *path)
     return filename;
 }
 
+/* Expects an empty buffer, and never resizes the buffer. */
+static inline void dirname_from_path(buffer dest, const char *path)
+{
+    int pathlen = runtime_strlen(path);
+    const char *last_delim = (char *)utf8_find_r((u8 *)path, '/');
+    const char *dirname;
+    int len;
+    if (!last_delim) {
+        dirname = path;
+        len = pathlen;
+    } else if (last_delim < path + pathlen - 1) {
+        dirname = last_delim + 1;
+        len = pathlen - (last_delim + 1 - path);
```

Comment:
A total nit, but "len = pathlen - (dirname - path)" seems clearer.

Reply:
Agreed. Also changed `last_delim - (delim + 1)` to `last_delim - dirname` a few lines below.